federation/schema.go

```diff
@@ -66,13 +67,55 @@ func validateFederationKeys(serviceNames []string, serviceSchemasByName map[stri
 						}
 					}
 				}
+				if isRootObject && !validFederatedKey {
+					return oops.Errorf("Invalid federation key %s", keyField)
+				}
+			}
+		}
+	}
+	return nil
+}
 
+// validateFederatedObjects validates that if a object is federated, it is federated on all the schemas
+func validateFederatedObjects(serviceNames []string, serviceSchemasByName map[string]*IntrospectionQueryResult, objName string) error {
+	// Check if it is federated on one service. It is federated if there is a field
+	//	called "__federation" on that object on any of the services it is on
+	federatedOnOneService := false
+	for _, service := range serviceNames {
+		for _, typ := range serviceSchemasByName[service].Schema.Types {
+			if typ.Name == objName {
+				for _, introspectedField := range typ.Fields {
+					if introspectedField.Name == federationField {
+						federatedOnOneService = true
+						break
+					}
+				}
 			}
+
 		}
 	}
-	if !validFederatedKey {
-		return oops.Errorf("Invalid federation key %s", keyField)
+
+	if federatedOnOneService && objName != "Query" && objName != "Mutation" {
+		// If the object is federated on one service, make sure every service with
+		// that object can parse the keys and fetch the object
+		for _, service := range serviceNames {
+			for _, typ := range serviceSchemasByName[service].Schema.Types {
+				if typ.Name == objName {
+					isRootObject := false
```

Comment:
is my understanding correct that the api change makes it so that root object doesn't really mean anything anymore; aka all federated objects are the same now with no root/shadow distinction? in that case should we call this `isFederatedObject` so it's less confusing?

Reply:
yup, renamed!